## 1. Dataset Understanding

# i. Load data & Preview Dataset dimensions

In [4]:
import pandas as pd
import numpy as np

In [5]:
df = pd.read_csv("data/orders.csv")

In [3]:
df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,1.0,29.99,8.72,38.71,1.0,credit_card,4.0
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00,1.0,118.70,22.76,141.46,1.0,boleto,4.0
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00,1.0,159.90,19.22,179.12,3.0,credit_card,5.0
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00,1.0,45.00,27.20,72.20,1.0,credit_card,5.0
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00,1.0,19.90,8.72,28.62,1.0,credit_card,5.0


In [4]:
df.tail()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
99436,9c5dedf39a927c1b2549525ed64a053c,39bd1228ee8140590ac3aca26f2dfe00,delivered,2017-03-09 09:54:05,2017-03-09 09:54:05,2017-03-10 11:18:03,2017-03-17 15:08:01,2017-03-28 00:00:00,1.0,72.00,13.08,85.08,3.0,credit_card,5.0
99437,63943bddc261676b46f01ca7ac2f7bd8,1fca14ff2861355f6e5f14306ff977a7,delivered,2018-02-06 12:58:58,2018-02-06 13:10:37,2018-02-07 23:22:42,2018-02-28 17:37:56,2018-03-02 00:00:00,1.0,174.90,20.10,195.00,3.0,credit_card,4.0
99438,83c1379a015df1e13d02aae0204711ab,1aa71eb042121263aafbe80c1b562c9c,delivered,2017-08-27 14:46:43,2017-08-27 15:04:16,2017-08-28 20:52:26,2017-09-21 11:24:17,2017-09-27 00:00:00,1.0,205.99,65.02,271.01,5.0,credit_card,5.0
99439,11c177c8e97725db2631073c19f07b62,b331b74b18dc79bcdf6532d51e1637c1,delivered,2018-01-08 21:28:27,2018-01-08 21:36:21,2018-01-12 15:35:03,2018-01-25 23:32:54,2018-02-15 00:00:00,2.0,359.98,81.18,441.16,4.0,credit_card,2.0
99440,66dea50a8b16d9b4dee7af250b4be1a5,edb027a75a1449115f6b43211ae02a24,delivered,2018-03-08 20:57:30,2018-03-09 11:20:28,2018-03-09 22:11:59,2018-03-16 13:08:30,2018-04-03 00:00:00,1.0,68.50,18.36,86.86,1.0,debit_card,5.0


In [5]:
df.shape

(99441, 15)

In [6]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 99441
Number of columns: 15


### Observation

The dataset contains 99441 records and 15 variables.
Each record represents an e-commerce order-level observation.

In [7]:
df.columns.tolist()

['order_id',
 'customer_id',
 'order_status',
 'order_purchase_timestamp',
 'order_approved_at',
 'order_delivered_carrier_date',
 'order_delivered_customer_date',
 'order_estimated_delivery_date',
 'n_items',
 'price',
 'freight',
 'payment',
 'installments',
 'payment_type',
 'review_score']

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 15 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   order_id                       99441 non-null  object 
 1   customer_id                    99441 non-null  object 
 2   order_status                   99441 non-null  object 
 3   order_purchase_timestamp       99441 non-null  object 
 4   order_approved_at              99281 non-null  object 
 5   order_delivered_carrier_date   97658 non-null  object 
 6   order_delivered_customer_date  96476 non-null  object 
 7   order_estimated_delivery_date  99441 non-null  object 
 8   n_items                        98666 non-null  float64
 9   price                          98666 non-null  float64
 10  freight                        98666 non-null  float64
 11  payment                        99440 non-null  float64
 12  installments                   99440 non-null 

## ii.. Column Classification

In [9]:
df.select_dtypes(include=np.number).columns.tolist()

['n_items', 'price', 'freight', 'payment', 'installments', 'review_score']

In [10]:
df.select_dtypes(include="object").columns.tolist()

['order_id',
 'customer_id',
 'order_status',
 'order_purchase_timestamp',
 'order_approved_at',
 'order_delivered_carrier_date',
 'order_delivered_customer_date',
 'order_estimated_delivery_date',
 'payment_type']

In [11]:
df.describe()

,n_items,price,freight,payment,installments,review_score
count,98666.000000,98666.000000,98666.000000,99440.000000,99440.000000,98673.000000
mean,1.141731,137.754076,22.871095,160.990267,2.930792,4.086793
std,0.538452,210.645145,21.825135,221.951257,2.715942,1.346274
min,1.000000,0.850000,0.000000,0.000000,0.000000,1.000000
25%,1.000000,45.900000,13.850000,62.010000,1.000000,4.000000
50%,1.000000,86.900000,17.180000,105.290000,2.000000,5.000000
75%,1.000000,149.900000,24.060000,176.970000,4.000000,5.000000
max,21.000000,13440.000000,1794.960000,13664.080000,24.000000,5.000000


In [12]:
df.describe(include="all")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
count,99441,99441,99441,99441,99281,97658,96476,99441,98666.000000,98666.000000,98666.000000,99440.000000,99440.000000,99440,98673.000000
unique,99441,99441,8,98875,90733,81018,95664,459,NaN,NaN,NaN,NaN,NaN,5,NaN
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-04-11 10:48:14,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-08 23:38:46,2017-12-20 00:00:00,NaN,NaN,NaN,NaN,NaN,credit_card,NaN
freq,1,1,96478,3,9,47,3,522,NaN,NaN,NaN,NaN,NaN,75387,NaN
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.141731,137.754076,22.871095,160.990267,2.930792,NaN,4.086793
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.538452,210.645145,21.825135,221.951257,2.715942,NaN,1.346274
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000,0.850000,0.000000,0.000000,0.000000,NaN,1.000000
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000,45.900000,13.850000,62.010000,1.000000,NaN,4.000000
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000,86.900000,17.180000,105.290000,2.000000,NaN,5.000000
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000,149.900000,24.060000,176.970000,4.000000,NaN,5.000000


In [13]:
missing = df.isnull().sum()

missing[missing > 0].sort_values(ascending=False)

order_delivered_customer_date    2965
order_delivered_carrier_date     1783
n_items                           775
price                             775
freight                           775
review_score                      768
order_approved_at                 160
payment                             1
installments                        1
payment_type                        1
dtype: int64

In [14]:
missing_percentage = (df.isnull().sum() / len(df) * 100)

missing_percentage[missing_percentage > 0].sort_values(ascending=False)

order_delivered_customer_date    2.981668
order_delivered_carrier_date     1.793023
n_items                          0.779357
price                            0.779357
freight                          0.779357
review_score                     0.772317
order_approved_at                0.160899
payment                          0.001006
installments                     0.001006
payment_type                     0.001006
dtype: float64

In [15]:
missing_report = pd.DataFrame({
    "Missing Values": df.isnull().sum(),
    "Missing Percentage": df.isnull().mean() * 100
})

missing_report = missing_report[
    missing_report["Missing Values"] > 0
].sort_values("Missing Values", ascending=False)

missing_report

,Missing Values,Missing Percentage
order_delivered_customer_date,2965,2.981668
order_delivered_carrier_date,1783,1.793023
n_items,775,0.779357
price,775,0.779357
freight,775,0.779357
review_score,768,0.772317
order_approved_at,160,0.160899
payment,1,0.001006
installments,1,0.001006
payment_type,1,0.001006


In [16]:
df.duplicated().sum()

0

In [17]:
df["order_id"].duplicated().sum()

0

In [18]:
df["order_status"].value_counts(dropna=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [19]:
df["payment_type"].value_counts(dropna=False)

payment_type
credit_card    75387
boleto         19784
voucher         2739
debit_card      1527
not_defined        3
NaN                1
Name: count, dtype: int64

In [20]:
df["review_score"].value_counts(dropna=False).sort_index()

review_score
1.000000    11316
1.500000        8
2.000000     3125
2.500000       34
3.000000     8136
3.333333        1
3.500000       25
4.000000    19018
4.333333        1
4.500000       54
5.000000    56955
NaN           768
Name: count, dtype: int64

In [21]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

df[date_columns].head()

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


## 3. Initial Observations

### Dataset Overview
- The dataset contains **99,441 order-level records and 15 variables** covering order details, customer identifiers, payment information, delivery timestamps, and review scores.
- `order_id` and `customer_id` are unique in this prepared dataset, indicating that each row represents a distinct order record.

### Data Quality
- No completely duplicated rows were identified.
- Missing values are present in several columns, with the highest proportion found in `order_delivered_customer_date` (approximately 2.98%).
- Missing delivery-related dates may be associated with orders that were not successfully delivered and will require further investigation before treatment.
- `n_items`, `price`, and `freight` each have approximately 0.78% missing values, while `review_score` has approximately 0.77% missing values.

### Order and Payment Overview
- The majority of orders have a `delivered` status, accounting for approximately 97% of the dataset.
- `credit_card` is the most frequently recorded payment method, followed by `boleto`.
- Review scores range from 1 to 5, providing an opportunity to study customer satisfaction in relation to order and delivery characteristics.

### Data Preparation Requirements
- The date-related columns are currently stored as non-datetime values and will require appropriate datetime conversion during preprocessing.
- Missing values will be investigated based on their business context rather than being removed or filled blindly.
- Numerical variables such as `price`, `freight`, and `payment` show wide ranges and will require further distribution and outlier analysis.

### Next Steps
- Investigate the causes and patterns of missing values.
- Convert and validate date/time columns.
- Perform data cleaning and preprocessing.
- Create useful analytical features such as delivery duration and delivery delay.
- Continue with purchase, payment, delivery, and customer satisfaction analysis.

# 2. Data Quality Analysis

## i. Missing Value Analysis

In [22]:
missing_report = pd.DataFrame({
    "Missing Values": df.isnull().sum(),
    "Missing Percentage": (df.isnull().mean() * 100).round(2)
})

missing_report = missing_report[
    missing_report["Missing Values"] > 0
].sort_values("Missing Values", ascending=False)

missing_report


,Missing Values,Missing Percentage
order_delivered_customer_date,2965,2.98
order_delivered_carrier_date,1783,1.79
n_items,775,0.78
price,775,0.78
freight,775,0.78
review_score,768,0.77
order_approved_at,160,0.16
payment,1,0.00
installments,1,0.00
payment_type,1,0.00


## ii. Duplicate Analysis

In [23]:
print("Complete duplicate rows:", df.duplicated().sum())
print("Duplicate order IDs:", df["order_id"].duplicated().sum())
print("Duplicate customer IDs:", df["customer_id"].duplicated().sum())

Complete duplicate rows: 0
Duplicate order IDs: 0
Duplicate customer IDs: 0


## iii. Categorical Value Analysis

In [24]:
df["order_status"].value_counts(dropna=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [25]:
df["payment_type"].value_counts(dropna=False)

payment_type
credit_card    75387
boleto         19784
voucher         2739
debit_card      1527
not_defined        3
NaN                1
Name: count, dtype: int64

In [26]:
df["review_score"].value_counts(dropna=False).sort_index()

review_score
1.000000    11316
1.500000        8
2.000000     3125
2.500000       34
3.000000     8136
3.333333        1
3.500000       25
4.000000    19018
4.333333        1
4.500000       54
5.000000    56955
NaN           768
Name: count, dtype: int64

## iv. Numerical Data Validation

In [27]:
numeric_columns = [
    "n_items",
    "price",
    "freight",
    "payment",
    "installments",
    "review_score"
]

df[numeric_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
n_items,98666.0,1.141731,0.538452,1.00,1.00,1.00,1.00,21.00
price,98666.0,137.754076,210.645145,0.85,45.90,86.90,149.90,13440.00
freight,98666.0,22.871095,21.825135,0.00,13.85,17.18,24.06,1794.96
payment,99440.0,160.990267,221.951257,0.00,62.01,105.29,176.97,13664.08
installments,99440.0,2.930792,2.715942,0.00,1.00,2.00,4.00,24.00
review_score,98673.0,4.086793,1.346274,1.00,4.00,5.00,5.00,5.00


In [28]:
for column in numeric_columns:
    print(f"{column}: {(df[column] < 0).sum()} negative values")

n_items: 0 negative values
price: 0 negative values
freight: 0 negative values
payment: 0 negative values
installments: 0 negative values
review_score: 0 negative values


In [29]:
invalid_reviews = df[
    (df["review_score"] < 1) |
    (df["review_score"] > 5)
]

print("Invalid review scores:", len(invalid_reviews))

Invalid review scores: 0


### v. Non-Integer Review Score Check

In [35]:
non_integer_reviews = df[
    df["review_score"].notna() &
    (df["review_score"] % 1 != 0)
]

print("Non-integer review scores:", len(non_integer_reviews))

Non-integer review scores: 123


In [36]:
non_integer_reviews["review_score"].value_counts().sort_index()

review_score
1.500000     8
2.500000    34
3.333333     1
3.500000    25
4.333333     1
4.500000    54
Name: count, dtype: int64

In [30]:
invalid_installments = df[df["installments"] < 1]

print("Invalid installment records:", len(invalid_installments))

Invalid installment records: 2


## vi. Missing Delivery Information

In [31]:
df[df["order_delivered_customer_date"].isna()]["order_status"].value_counts(dropna=False)

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

In [32]:
df[df["order_delivered_carrier_date"].isna()]["order_status"].value_counts(dropna=False)

order_status
unavailable    609
canceled       550
invoiced       314
processing     301
created          5
approved         2
delivered        2
Name: count, dtype: int64

## vii. Date Column Validation

In [33]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

df[date_columns].dtypes

order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

In [34]:
df[
    [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
].head(10)

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00
5,2017-07-09 21:57:05,2017-07-09 22:10:13,2017-07-11 14:58:04,2017-07-26 10:57:55,2017-08-01 00:00:00
6,2017-04-11 12:22:08,2017-04-13 13:25:17,NaN,NaN,2017-05-09 00:00:00
7,2017-05-16 13:10:30,2017-05-16 13:22:11,2017-05-22 10:07:46,2017-05-26 12:55:51,2017-06-07 00:00:00
8,2017-01-23 18:29:09,2017-01-25 02:50:47,2017-01-26 14:16:31,2017-02-02 14:08:10,2017-03-06 00:00:00
9,2017-07-29 11:55:02,2017-07-29 12:05:32,2017-08-10 19:45:24,2017-08-16 17:14:30,2017-08-23 00:00:00


### viii. fractional reviews

In [37]:
non_integer_reviews = df[
    df["review_score"].notna() &
    (df["review_score"] % 1 != 0)
]

print("Non-integer review scores:", len(non_integer_reviews))

Non-integer review scores: 123


In [38]:
non_integer_reviews["review_score"].value_counts().sort_index()

review_score
1.500000     8
2.500000    34
3.333333     1
3.500000    25
4.333333     1
4.500000    54
Name: count, dtype: int64

In [39]:
invalid_installments

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
63782,744bade1fcf9ff3f31d860ace076d422,5e5794daaa13f73e2f1cdb4114529843,delivered,2018-04-22 11:34:42,2018-04-24 19:04:46,2018-04-24 03:14:34,2018-04-27 20:55:28,2018-05-16 00:00:00,1.0,45.90,12.79,58.69,0.0,credit_card,5.0
66368,1a57108394169c0b47d8f876acc9ba2d,48ebb06cf56dba9d009230cc751bb195,delivered,2018-05-15 16:25:14,2018-05-15 16:36:52,2018-05-17 12:37:00,2018-05-24 15:45:41,2018-06-06 00:00:00,2.0,83.38,46.56,129.94,0.0,credit_card,5.0


## ix. Data Quality Observations

### Missing Values
- Missing values are present in several columns, with the highest proportion found in `order_delivered_customer_date` at approximately 2.98%.
- `order_delivered_carrier_date` has approximately 1.79% missing values, while `n_items`, `price`, `freight`, and `review_score` each have less than 1% missing values.
- Missing delivery dates are largely associated with orders that have not completed the corresponding delivery stage.

### Duplicate Records
- No completely duplicated rows were identified.
- No duplicate `order_id` values were identified.
- No duplicate `customer_id` values were identified.
- Since `customer_id` is unique in this prepared dataset, repeat-customer behavior should not be inferred directly from this dataset.

### Categorical Data
- Order status values represent different stages of the order lifecycle.
- `credit_card` is the most common payment method, followed by `boleto`.
- Only a very small number of records contain `not_defined` or missing payment types.

### Numerical Data
- No negative values were found in the checked numerical variables.
- `price`, `freight`, and `payment` have wide ranges, indicating that their distributions and potential outliers should be investigated during exploratory analysis.
- Two records contain an installment value of `0` and require further investigation before deciding how they should be treated.

### Review Scores
- `review_score` contains 768 missing values.
- All recorded review scores fall within the range of 1 to 5.
- 123 records contain non-integer review scores, including values such as `1.5`, `2.5`, `3.5`, `4.5`, `3.333333`, and `4.333333`.
- These fractional review scores require further investigation before deciding whether any preprocessing or transformation is necessary.

### Delivery Data
- Most missing delivery dates are consistent with orders that have not reached the corresponding delivery stage.
- However, a small number of orders marked as `delivered` have missing delivery dates and should be investigated separately.

### Date Columns
- All five date-related columns are currently stored as `object` data types.
- These columns will require datetime conversion during the preprocessing stage.

### Conclusion
The dataset does not contain duplicate order records or negative numerical values, but several data-quality aspects require further investigation. These include missing values, fractional review scores, zero installment values, and missing delivery dates for a small number of delivered orders. These issues will be evaluated during the data cleaning and preprocessing stages before performing detailed analysis.

# 3. Data Cleaning & Preprocessing

## 3.1 Create a Clean Dataset

In [6]:
df_clean = df.copy()

print("Original dataset shape:", df.shape)
print("Clean dataset shape:", df_clean.shape)

Original dataset shape: (99441, 15)
Clean dataset shape: (99441, 15)


In [7]:
before = len(df_clean)

df_clean = df_clean.drop_duplicates()

after = len(df_clean)

print("Rows before removing duplicates:", before)
print("Rows after removing duplicates:", after)
print("Rows removed:", before - after)

Rows before removing duplicates: 99441
Rows after removing duplicates: 99441
Rows removed: 0


In [8]:
print("Unique order IDs:", df_clean["order_id"].nunique())
print("Total rows:", len(df_clean))

Unique order IDs: 99441
Total rows: 99441


## 3.2 Standardize Categorical Values

In [9]:
categorical_columns = [
    "order_status",
    "payment_type"
]

for column in categorical_columns:
    df_clean[column] = df_clean[column].str.strip().str.lower()

In [10]:
for column in categorical_columns:
    print(f"\n{column}:")
    print(df_clean[column].value_counts(dropna=False))


order_status:
order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

payment_type:
payment_type
credit_card    75387
boleto         19784
voucher         2739
debit_card      1527
not_defined        3
NaN                1
Name: count, dtype: int64


## 3.3 Convert Date Columns

In [11]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for column in date_columns:
    df_clean[column] = pd.to_datetime(
        df_clean[column],
        errors="coerce"
    )

In [12]:
df_clean[date_columns].dtypes

order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object

In [13]:
date_missing = df_clean[date_columns].isnull().sum()

date_missing

order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [14]:
df_clean[
    df_clean["price"].isna()
]["order_status"].value_counts(dropna=False)

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [15]:
df_clean[
    df_clean["n_items"].isna()
]["order_status"].value_counts(dropna=False)

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [16]:
df_clean[
    df_clean["freight"].isna()
]["order_status"].value_counts(dropna=False)

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [17]:
df_clean["review_score"].isna().sum()

768

In [18]:
df_clean[df_clean["installments"] == 0]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
63782,744bade1fcf9ff3f31d860ace076d422,5e5794daaa13f73e2f1cdb4114529843,delivered,2018-04-22 11:34:42,2018-04-24 19:04:46,2018-04-24 03:14:34,2018-04-27 20:55:28,2018-05-16,1.0,45.90,12.79,58.69,0.0,credit_card,5.0
66368,1a57108394169c0b47d8f876acc9ba2d,48ebb06cf56dba9d009230cc751bb195,delivered,2018-05-15 16:25:14,2018-05-15 16:36:52,2018-05-17 12:37:00,2018-05-24 15:45:41,2018-06-06,2.0,83.38,46.56,129.94,0.0,credit_card,5.0


In [19]:
df_clean.shape

(99441, 15)

In [20]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 15 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
 8   n_items                        98666 non-null  float64       
 9   price                          98666 non-null  float64       
 10  freight                        98666 non-null  float64       
 11  payment        

In [21]:
df_clean.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,1.0,29.99,8.72,38.71,1.0,credit_card,4.0
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,1.0,118.70,22.76,141.46,1.0,boleto,4.0
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,1.0,159.90,19.22,179.12,3.0,credit_card,5.0
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,1.0,45.00,27.20,72.20,1.0,credit_card,5.0
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,1.0,19.90,8.72,28.62,1.0,credit_card,5.0


In [22]:
df_clean[df_clean["price"].isna()]["order_status"].value_counts(dropna=False)

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [23]:
df_clean[df_clean["n_items"].isna()]["order_status"].value_counts(dropna=False)

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [24]:
df_clean[df_clean["freight"].isna()]["order_status"].value_counts(dropna=False)

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [25]:
df_clean[df_clean["installments"] == 0]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
63782,744bade1fcf9ff3f31d860ace076d422,5e5794daaa13f73e2f1cdb4114529843,delivered,2018-04-22 11:34:42,2018-04-24 19:04:46,2018-04-24 03:14:34,2018-04-27 20:55:28,2018-05-16,1.0,45.90,12.79,58.69,0.0,credit_card,5.0
66368,1a57108394169c0b47d8f876acc9ba2d,48ebb06cf56dba9d009230cc751bb195,delivered,2018-05-15 16:25:14,2018-05-15 16:36:52,2018-05-17 12:37:00,2018-05-24 15:45:41,2018-06-06,2.0,83.38,46.56,129.94,0.0,credit_card,5.0


### 3.4 Investigate Delivered Orders with Missing Delivery Dates

In [26]:
delivered_missing_customer_date = df_clean[
    (df_clean["order_status"] == "delivered") &
    (df_clean["order_delivered_customer_date"].isna())
]

print(
    "Delivered orders with missing customer delivery date:",
    len(delivered_missing_customer_date)
)

Delivered orders with missing customer delivery date: 8


In [27]:
delivered_missing_carrier_date = df_clean[
    (df_clean["order_status"] == "delivered") &
    (df_clean["order_delivered_carrier_date"].isna())
]

print(
    "Delivered orders with missing carrier date:",
    len(delivered_missing_carrier_date)
)

Delivered orders with missing carrier date: 2


In [28]:
delivered_missing_customer_date

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18,1.0,117.30,17.53,134.83,3.0,credit_card,5.0
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16,1.0,329.00,25.24,354.24,1.0,debit_card,5.0
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaT,2018-07-30,1.0,139.00,19.07,158.07,3.0,credit_card,5.0
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaT,2018-07-30,1.0,139.00,19.07,158.07,1.0,credit_card,5.0
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaT,2018-07-24,1.0,188.99,15.63,204.62,4.0,credit_card,5.0
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23,1.0,179.00,15.00,194.00,4.0,credit_card,5.0
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaT,2018-06-26,1.0,110.99,9.13,120.12,5.0,credit_card,1.0
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaT,2018-07-19,1.0,45.90,9.07,54.97,1.0,credit_card,5.0


In [29]:
delivered_missing_carrier_date

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
73222,2aa91108853cecb43c84a5dc5b277475,afeb16c7f46396c0ed54acb45ccaaa40,delivered,2017-09-29 08:52:58,2017-09-29 09:07:16,NaT,2017-11-20 19:44:47,2017-11-14,1.0,179.0,14.98,193.98,1.0,credit_card,5.0
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23,1.0,179.0,15.00,194.00,4.0,credit_card,5.0


## 3.5 Data Cleaning & Preprocessing Observations

- Created a separate `df_clean` dataset while preserving the original data.
- No duplicate records were found or removed.
- Categorical values were standardized and all date columns were converted to datetime format.
- Missing `price`, `n_items`, and `freight` values were mainly associated with unavailable or canceled orders.
- Missing reviews, fractional review scores, and zero installment values were retained for further investigation.
- A small number of delivered orders were found with missing delivery dates and were retained as data-quality anomalies.

### Conclusion

The dataset has been cleaned and standardized while preserving potentially meaningful records. It is now ready for feature engineering and further analysis.